### PR TITLE
Add opt-in environment lifetime cancellation

### PR DESCRIPTION
**Pull request by djgilcrease**
Connection setup and socket writes can outlive the configured RPC timeout, and application shutdown cannot interrupt locator reconnection sleeps. This PR adds `NewEnvironmentWithContext(ctx, options)`, which opts into cancellation for the whole environment lifetime, including every producer and consumer connection.

The context interrupts:
- TCP dialing
- TLS negotiation
- advertised-host DNS lookup
- socket I/O
- protocol response waits
- locator and advertised-endpoint retry delays

Closing an environment created with this constructor cancels its lifetime before releasing resources, and concurrent closes share one cleanup owner. Existing `NewEnvironment` callers keep graceful-close behaviour and their configured RPC timeouts. Caller options are copied, so each environment can have its own lifetime.

The socket cancellation callback closes the underlying connection without taking the socket writer's mutex.

The README explains:
- lifetime contexts versus construction-only contexts
- checking for cancellation with `errors.Is`
- the outcome of pending publications
- explicitly shutting down the HA wrappers

This change covers `pkg/stream`. It does not add cancellation to application callbacks, or to the separate retry timers and send waits in `pkg/ha`.

## Split-out fixes
As requested in review, fixes for problems that already exist on `main` are now separate PRs:
- #527 return RPC write errors immediately; discard abandoned responses without closing their channels
- #528 infer TLS server name from the connection host
- #529 fresh client for each bootstrap seed, plus the flaky metrics publish-error spec
- #530 close locator connections and stop reconnecting after `Environment.Close`, plus the locator client race
- #531 `StreamMetaData` leader retry nil panic

This PR is now a single feature commit. Two of its specs will fail until their split-out fixes merge:
- "stops the leader retry loop" needs #531.
- "discards a response abandoned by an RPC timeout on a lifetime client" needs #527.

I'll rebase once those land. #529 and #530 also touch `NewEnvironment` and the locator, so expect a small rebase after they merge.

## Validation
- The full randomized, parallel Ginkgo suites run with `--race --procs=2` against an isolated RabbitMQ broker. Every spec passes except the two listed above.
- The cancellation specs cover:
  - a stalled dial
  - stalled TLS, write and protocol handshakes
  - code and data response waits, and their cleanup
  - advertised-host DNS lookup
  - metadata leader retries
  - reconnect/close interruption
  - active producer shutdown
  - reuse of caller options
- `golangci-lint` reports nothing new; its only findings already exist on `main`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_014f1pquAiUfw5incLoLQChg

### CODE REVIEW

**Comment by Gsantomaggio** [comment]
Hi @djgilcrease 👋🏽
Thank you for your contribution

To move forward with the PR, you should sign a CLA.
You can find the process here: https://github.com/rabbitmq/cla

**Comment by djgilcrease** [comment]
signed the CLA

**Comment by Gsantomaggio** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by Gsantomaggio** [comment]
@djgilcrease Thank you for the PR. It is useful.
I will have a look on it soon, in the meantime I asked a review to the AI. Can you please have a look on these comments ?



---

pkg/stream/socket.go line 96-97 (in handleWriteWithResponse)

▎ Potential panic: send on closed channel for plain RPC timeouts.
▎
▎ This only skips RemoveResponseById when connectionContext().Err() != nil (line 91), but waitCode's plain timeout branch (context.go:84-85) returns an error without canceling the context — a normal, unrelated-to-cancellation RPC timeout. In that case we still fall through to RemoveResponseById, which does close(resp.code) / close(resp.data) (coordinator.go:118-119), while a frame-reader goroutine may still be about to res.code <- ... / res.data <- ... unconditionally (server_frame.go, many sites, e.g. lines 179, 213, 241...). That's a send on closed channel panic.
▎
▎ discardResponse was added specifically to avoid this hazard for the cancellation path — should the plain-timeout branch use it too, instead of RemoveResponseById?

---

pkg/stream/environment.go line 381-387 (in Close)

▎ Data race on env.locator.client for plain NewEnvironment.
▎
▎ The env.locator.mutex.Lock() guarding the read of env.locator.client is only taken if env.cancel != nil — i.e. only for environments created via NewEnvironmentWithContext. Plain stream.NewEnvironment() environments get no lock here, and Close() can race with maybeReconnectLocator's locked writes to env.locator.client (e.g. a concurrent DeclareStream call). Under -race this will flag, and in the worst case Close() can close a half-constructed client or miss closing the client that replaced it.
▎
▎ Since taking the lock is cheap, could this just be unconditional (drop the if env.cancel != nil guard on the lock itself)?

---

pkg/stream/client.go line 787-790 (in BrokerLeaderWithResolver)

▎ Discards a valid/known result in favor of an unrelated cancellation error.
▎
▎ This check is new in this PR. After lookup(ctx, ...) returns — success or failure — the code unconditionally checks the parent connectionContext().Err() and returns that instead. Two problems:
▎ 1. If the environment happens to start closing while the DNS lookup is in flight, a successful lookup result gets thrown away and replaced with an unrelated context-canceled error.
▎ 2. If the lookup itself fails with a real DNS error at the same time, that error is silently swallowed instead of being logged (as the existing dnsError branch below does).
▎
▎ Suggest only checking the context error as a substitute when lookup itself errors, e.g. if err != nil { if ctxErr := ...; ctxErr != nil { return nil, ctxErr } }, rather than short-circuiting on every call regardless of outcome.

---

pkg/stream/context.go line 84-85 (in waitCode, timeout branch) — minor

▎ The context-aware timeout branch drops the logs.LogError diagnostic that waitCodeWithTimeOut (utils.go:49) still emits for the identical timeout condition. Worth adding back for parity so ordinary timeouts on NewEnvironmentWithContext clients aren't silently harder to diagnose than on plain clients.

---

pkg/stream/client.go connect() (~line 183 area) — minor

▎ The dial and TCP-tuning failure paths (SetWriteBuffer/SetReadBuffer/SetNoDelay) no longer log via logs.LogDebug/logs.LogError before returning, unlike the pre-diff code. Was dropping these intentional, or an oversight during the refactor?

---

pkg/stream/environment.go line ~178 (in newEnvironment) — minor

▎ client.Close() is called explicitly here in the failure branch, and then unconditionally again via the trailing defer. This relies on Close()'s idempotency rather than closing exactly once — worth a comment noting that's intentional, or restructuring to close once.

---

**Comment by djgilcrease** [comment]
Thanks for the review! I've rebased on main and addressed everything in separate commits:
1. Send on closed channel after a timeout: confirmed. The race already existed on main: the frame reader looks up the response, releases the lock, then sends. Any waitCode error now returns early and the deferred discardResponse removes the response without closing its channels. They're buffered for the one value each handler sends and nothing ranges over them, so nothing blocks or leaks. Client.Close had the same window with its close response and is fixed too. The new specs panicked before the fix.
2. Race on env.locator.client: confirmed, and it was also on main. Every environment operation read the field without the lock after maybeReconnectLocator returned, not just Close. I didn't make the lock unconditional: on a plain NewEnvironment, the reconnect loop holds that mutex and retries for as long as the brokers are unreachable, so Close would hang. Instead locator.client is now an atomic.Pointer[Client]. The mutex only serializes reconnects, plus the bounded wait in Close for context environments, which cancel first. A new spec reproduces the race under -race and passes now.
3. BrokerLeaderWithResolver: agreed. The context is only checked when the lookup fails, and the lookup error is joined with the cancellation, so errors.Is(err, context.Canceled) still holds and the DNS error isn't lost.
4. Timeout log in waitCode: restored. I also fixed the existing message, which labeled milliseconds as "ns".
5. connect() logs: that was an oversight. The dial and SetWriteBuffer/SetReadBuffer/SetNoDelay logs are back with their original messages. The dial log also covers the old resolve-error log, since DialContext wraps the DNS error. It's skipped when the dial was interrupted by lifetime cancellation.
6. Double client.Close(): each seed attempt is now closed exactly once inside the loop, and the trailing defer is gone.

The full suite passes with --race --procs=2.

While investigating and fixing these, I found a potential issue that already exists on main. For an environment created with plain NewEnvironment, the reconnect loop never closes the clients it replaces. It can also install a new client after Close has run, so that client leaks. To fix it, the loop has to notice Close. Should I open a separate PR to fix this?

**Comment by Gsantomaggio** [comment]
Thank you @djgilcrease,
For everything that is in the `main` we can open another PR(s). This PR is already big :).

Let's focus on this PR.

**Comment by djgilcrease** [comment]
Thanks! I've split the fixes for problems that were already on main into their own PRs: #527 (RPC write errors and abandoned responses), #528 (TLS server name), #529 (bootstrap seeds), #530 (locator reconnect/close and the locator race), and #531 (StreamMetaData retry nil panic). This PR is now a single commit containing only the context-cancellation feature.

 Two specs here depend on #527 and #531, so they'll fail until those merge. I'll rebase once they land.

**Comment by Gsantomaggio** [comment]
Thank you @djgilcrease,
Can you provide the PR(s)review order? So I can review, merge, and move forward?

**Comment by djgilcrease** [comment]
> Thank you @djgilcrease, Can you provide the PR(s)review order? So I can review, merge, and move forward?

1. #529 (seed isolation plus the metrics spec fix): has no code conflicts with any of the others. Landing it early also stops the flaky metrics spec from failing the remaining PRs' CI.
2. #527 (RPC abandoned responses): #523's timeout spec needs it. Its only code conflict is with #523 (socket.go).
3. #528 (TLS server name): only conflicts in code with #523 (client.go).
4. #531 (StreamMetaData nil guard): a one-line fix. Its only code conflict is with #530 (the same loop in environment.go), so landing it first means #530's rebase is trivial.
5. #530 (locator reconnect/close): the biggest change, and the behaviour change (AlreadyClosed after Close) needs a real review. It conflicts with #531 and #523 in environment.go, so merge it after #531 and before #523.
6. #523 (feature): last. It has code conflicts with #527, #528 and #530. Rebasing once onto all of them resolves everything together, and its two failing specs pass once #527 and #531 are in.

**Comment by Gsantomaggio** [comment]
@djgilcrease Thank you.
FYI: I have added https://github.com/rabbitmq/rabbitmq-stream-go-client/tree/main/examples/reliable. This is a long running test I usually use to validate the library for the autoreconnection part.  

We have different long running tests that we can't run on the CI. 
Please run it before propose a PR to see if everything is ok.

**Comment by djgilcrease** [comment]
@Gsantomaggio Thanks for the pointer — I rebased this branch onto current `main` (it now includes #529 and your #533 scripts, no conflicts) and ran `examples/reliable` as you asked.

I ran it **twice**: once exactly as it is on `main` (`NewEnvironment`), and once with a local copy switched to `NewEnvironmentWithContext` so this PR's new code path gets the same treatment. Only two constants were changed, identically in both, because the defaults finish all 14M messages in under a minute and no chaos script ever gets a chance to interfere: `messagesToSend = 2_000_000` and `sendDelay = 1 * time.Second`, giving a 28M-message target over ~1h50m each.

Setup: 3-node cluster + HAProxy from `compose/ha_tls`. Each run closed all connections (`close_all_connections`, 39 iterations) until half the messages were sent, then restarted containers (`restart_containers`, ~70 restarts) for the rest, as the README suggests not running both at once.

**`NewEnvironment` (unchanged example)** — 1h50m

```
MESSAGES
  Target  28,000,000      Sent  28,000,000      ReSent  0

  CONFIRMATIONS
  Confirmed  27,998,058    Failed  1,942         Total  28,000,000

  CONSUMPTION
  Total  55,996,910      Per Consumer  27,998,455
```

**`NewEnvironmentWithContext` (this PR's new constructor)** — 1h52m

```
MESSAGES
  Target  28,000,000      Sent  28,000,000      ReSent  0

  CONFIRMATIONS
  Confirmed  27,998,819    Failed  1,181         Total  28,000,000

  CONSUMPTION
  Total  55,999,292      Per Consumer  27,999,646
```

In both runs: `ReSent` is 0, `Confirmed + Failed` is exactly the target, and consumption is exactly twice the stored message count reported by the management API (2 consumers per stream), so no message was lost or double-counted. No panics, all 16 producers/consumers back to `Open` after every disruption, goroutine count flat (~110-120 throughout, from the same 5 at startup), and a clean `Close` of producers, consumers and environment at the end. A no-chaos baseline run also passed (14M sent = 14M confirmed, 0 failed).

The runs also covered two unplanned cases, both of which the library handled: a full cluster outage of ~3 minutes (all four containers down, everything reconnected on its own afterwards), and a single node being force-restarted during the final drain.

Two pre-existing issues showed up while running this. Both reproduce identically on unmodified `main` with `NewEnvironment`, so neither belongs to this PR:

1. **`Reliable*.Close()` during a reconnect returns `AlreadyClosed`.** If the example is closed while producers/consumers are still `Reconnecting` (e.g. right after a broker restart), `producer.Close()` returns `AlreadyClosed` from the inner entity that the dropped connection had already closed, and `CheckErr` exits. Worse, `Close` does not stop a reconnection in flight: for `ReliableProducer`/`ReliableSuperStreamProducer` the retry then creates a new producer and sets the status back to `Open`, leaving a live producer and TCP connection behind a closed client; `ReliableConsumer` re-subscribes and can deliver messages after `Close`. `SuperStreamProducer.Close` also returns on the first partition error, leaving the remaining partitions unclosed. This predates #522 (v1.7.0 and v1.8.3 behave the same); #522 fixed exactly this for `ReliableSuperStreamConsumer` only. I have a fix ready following that same pattern and will open a separate PR for it.
2. **Closing connections while the example is still starting up kills it.** `ha.NewReliableConsumer`/`NewReliableProducer` do not retry their first `newConsumer()`/`newProducer()`, so a `close_all_connections` that lands during setup makes the example exit with `timeout 10000 ms - waiting Code, operation: CommandSubscribe`. Easy to avoid by starting the chaos scripts only once everything is `Open`, but it might be worth retrying the initial connection, or at least noting it in the README.

Happy to run any other long-running scenario you use before this is merged.

File: pkg/stream/client.go [hunk 1]
```diff
@@ -350,6 +374,7 @@ func (c *Client) sendSaslAuthenticate(saslMechanism string, challengeResponse []
 	length := 2 + 2 + 4 + 2 + len(saslMechanism) + 4 + len(challengeResponse)
 	resp := c.coordinator.NewResponse(commandSaslAuthenticate)
 	respTune := c.coordinator.NewResponseWithName("tune")
+	defer c.coordinator.discardResponse(respTune)
```
Comment (hiimjako) [inline pkg/stream/client.go:378]: Could we handle the `GetResponseByName("tune")` error in `handleTune` before dereferencing `res`?

Cancellation can make this defer remove `respTune` while the reader is already processing a TUNE frame. In that race, the lookup returns an error and `res` is nil, but `handleTune` still sends to `res.data`, causing a nil-pointer panic.

I think the minimal fix would be something like:

```go
res, err := c.coordinator.GetResponseByName("tune")
if err != nil {
    logErrorCommand(err, "handleTune")
    return nil
}
```

Could you check whether returning `nil` is the right behavior for the current handler contract, and add a regression test where cancellation wins just before a late TUNE frame is dispatched?

File: pkg/stream/client.go [hunk 2]
```diff
@@ -688,7 +722,10 @@ func (c *Client) metaData(streams ...string) *StreamsMetadata {
 		return nil
 	}
 
-	data := <-resp.data
+	data, dataErr := c.waitData(resp)
+	if dataErr != nil {
+		return nil
```
Comment (hiimjako) [inline pkg/stream/client.go:735]: Could the `Environment.StreamExists` path preserve this error instead of turning it into `(false, nil)`?

If the lifetime is canceled after `maybeReconnectLocator` succeeds but while this metadata RPC is waiting, `Client.StreamExists` converts the nil metadata response to `false`, and `Environment.StreamExists` returns a nil error. Callers then cannot distinguish cancellation from a stream that does not exist.

A regression test that cancels during the metadata request would help keep `errors.Is(err, context.Canceled)` consistent with the rest of the new API.

File: pkg/stream/environment.go [hunk 3]
```diff
@@ -305,8 +369,23 @@ func (env *Environment) NewSuperStreamProducer(superStream string, superStreamPr
 }
 
 func (env *Environment) Close() error {
+	if env.cancel != nil {
+		if !env.closeStarted.CompareAndSwap(false, true) {
```
Comment (hiimjako) [inline pkg/stream/environment.go:382]: Having a single cleanup owner makes sense. Could we add a regression test with several goroutines calling `Close` while a locator reconnect or RPC is blocked? The test should verify that every caller returns only after the cleanup owner has finished.

There is also a separate state race here: `closed` is written by `Close` and read by `IsClosed` as an unsynchronized `bool`. To cover that under `-race`, the test would need to call `IsClosed` concurrently with `Close`, or `closed` could be made atomic.

File: pkg/stream/environment.go [hunk 4]
```diff
@@ -305,8 +369,23 @@ func (env *Environment) NewSuperStreamProducer(superStream string, superStreamPr
...
+			<-env.closeDone
+			return nil
+		}
+		defer close(env.closeDone)
+		env.cancel()
```
Comment (hiimjako) [inline pkg/stream/environment.go:387]: Could we avoid warning-level logs for close errors caused by this intentional lifetime cancellation?

Canceling here can close the socket before the subsequent graceful close frame is written. The integration test already emits `use of closed network connection` warnings on this normal path.

One option would be to skip the close frame, or downgrade the error, only when the connection context is already canceled. I would avoid suppressing `net.ErrClosed` unconditionally, since it can still indicate an unexpected disconnect. The close warning in `pkg/stream/socket.go` has the same issue.

File: pkg/stream/brokers.go [hunk 5]
```diff
@@ -15,6 +17,9 @@ type AddressResolver struct {
 }
 
 type TCPParameters struct {
+	connectionContext     context.Context
```
Comment (hiimjako) [inline pkg/stream/brokers.go:20]: Not a blocker, but I would consider moving the lifetime context out of `TCPParameters` in a follow-up. It is runtime ownership state rather than TCP configuration.

Concretely, I would put a private `environmentLifecycle` in something like `pkg/stream/lifecycle.go`, owned by `Environment`. It could hold the context, cancel function, shutdown policy, and concurrent-close state.

Each `Client` should receive only the lifecycle context through `connectionParameters`, probably in `pkg/stream/pool.go`, rather than depending on the whole lifecycle. This would keep `TCPParameters` focused on TCP configuration and allow both constructors to share one context-aware implementation of `waitCode`, `waitData`, and `waitDelay`. `NewEnvironment` could use `context.Background()` to preserve its current behavior.

**Review by hiimjako — COMMENTED** [review]
Thanks for the detailed PR and for splitting the pre-existing fixes into separate changes. I like the direction here: an environment lifetime context is the right abstraction, and using `context.AfterFunc` to close the underlying connection without taking the writer mutex is a solid way to interrupt blocked I/O.

I would hold off on merging until #527, #530 and #531 are in and this branch has been rebased. The current branch still has the two acknowledged test panics. I also found two cancellation edge cases that those follow-up PRs do not cover: a late TUNE frame can panic after its response has been discarded, and `StreamExists` can turn cancellation into `(false, nil)`.

I do not think this needs another split. Most of the size is tests and documentation, and the production changes form one coherent feature.

File: pkg/stream/client.go (hunk 1 above)
Comment (djgilcrease) [reply pkg/stream/client.go:378]: Confirmed, and fixed as you suggested. `handleTune` was the only handler in `server_frame.go` that called `logErrorCommand` without returning — the eleven others all use the `if err != nil { logErrorCommand(...); return }` shape, so `return nil` matches the contract: the `any` result is discarded at the only call site, inside the fire-and-forget `handleResponse` switch.

Added `Abandoned response discarding [It] ignores a TUNE frame that arrives after its response was discarded`, which discards the `"tune"` response and then dispatches a TUNE frame. It panics without the guard. I drive `handleTune` directly rather than through `handleResponse` because a panic on the frame-reader goroutine cannot be caught by Ginkgo.

File: pkg/stream/client.go (hunk 2 above)
Comment (djgilcrease) [reply pkg/stream/client.go:735]: Confirmed. `Environment.StreamExists` now returns the lifetime error when the lookup came back false and the context is canceled, so `errors.Is(err, context.Canceled)` holds as it does for `StreamMetaData` and `BrokerLeaderWithResolver`. I fixed it at the `Environment` layer because `StreamExists(string) bool` is part of the exported `IClient` interface. It reuses the existing nil-safe `connectionContextError()`, so `NewEnvironment` behaviour is unchanged.

Regression spec: `Metadata lookup cancellation [It] reports cancellation instead of a missing stream`, which cancels only once the metadata request is on the wire so the result cannot come from the early check in `maybeReconnectLocator`.

Separately, while looking at this: `client.go:817` dereferences `streamsMetadata.Get(stream)`, which returns nil when the stream is absent from the reply. That one is pre-existing on `main`, so I'll send it as its own PR.

File: pkg/stream/environment.go (hunk 3 above)
Comment (djgilcrease) [reply pkg/stream/environment.go:382]: Both points are already fixed in #530, which is open: it makes `closed` an `atomic.Bool`, removes the locator-client race, and adds a spec asserting `Close` returns while a reconnect is blocked.

I've kept it there rather than duplicating it here, following the split Michael asked for — everything already on `main` goes in its own PR. This branch will pick it up on the rebase after #530 merges.

File: pkg/stream/environment.go (hunk 4 above)
Comment (djgilcrease) [reply pkg/stream/environment.go:387]: Confirmed and fixed at both sites, gated on cancellation rather than suppressing `net.ErrClosed` unconditionally.

`Client.Close` now skips the close frame entirely when the lifetime is already canceled. `socket.shutdown` downgrades to debug only when the error is `net.ErrClosed` **and** the lifetime is canceled, so an unexpected disconnect still warns. I applied the same guard to the `handleWrite` warning, which was noisy for the same reason.

The lifetime specs' captured output is now clean. The warnings left in the suite come from `environment_test.go` and `seed_recovery_test.go`, where the sockets really did break.

File: pkg/stream/brokers.go (hunk 5 above)
Comment (djgilcrease) [reply pkg/stream/brokers.go:20]: Agreed on both the diagnosis and that it's a follow-up. `TCPParameters` is TCP configuration and the lifetime is ownership state.

I'll open it once this lands: a private `environmentLifecycle` owned by `Environment`, with only the context reaching each `Client` through `connectionParameters`, which also lets `waitCode`/`waitData`/`waitDelay` drop their dual legacy branches.

**Comment by djgilcrease** [comment]
Thanks @hiimjako — all five comments are addressed, with replies in each thread.

Fixed here: the late-TUNE nil-pointer panic and `StreamExists` swallowing cancellation, each with a regression spec that fails without its fix. The close-error warnings are downgraded at both sites, gated on the lifetime actually being canceled so an unexpected disconnect still warns.

Not fixed here, deliberately: the `closed` race and the concurrent-`Close` test are already in #530, and the `environmentLifecycle` refactor will follow once this lands.

Rebased onto current `main`, which now has #527 — the `discards a response abandoned by an RPC timeout on a lifetime client` spec is green. #530 and #531 are still open, so `stops the leader retry loop` stays red until #531 lands, and I'll rebase again once both are in.

Full suite with `--race --procs=2 --randomize-all`: 265/266, the one failure being that known dependency. `golangci-lint` reports nothing new.
